authui/src/authflowv2/form.ts

```diff
@@ -0,0 +1,144 @@
+    e.preventDefault();
+  };
+
+  onSubmit = (e: Event) => {
+    this.submitForm(e);
+  };
+
+  async submitForm(e: Event) {
+    const form = e.currentTarget as HTMLFormElement;
+
+    if (form.querySelector('[data-turbo="false"]')) {
+      return;
+    }
+
+    // Do not stop propagation so that GTM can recognize the event as Form Submission trigger.
+    // e.stopPropagation();
+
+    const formData = new FormData(form);
+
+    const params = new URLSearchParams();
+    formData.forEach((value, name) => {
+      params.set(name, value as string);
+    });
+    const loadingEvent = new CustomEvent<{ button?: HTMLButtonElement }>(
+      "onLoading",
+      {
+        detail: {},
+      }
+    );
+    // FormData does not include any submit button's data:
+    // include them manually, since we have at most one submit button per form.
+    const submitButtons = form.querySelectorAll('button[type="submit"]');
+    for (let i = 0; i < submitButtons.length; i++) {
+      const button = submitButtons[i] as HTMLButtonElement;
+      params.set(button.name, button.value);
+    }
+    if (form.id) {
+      const el = document.querySelector(
+        `button[type="submit"][form="${form.id}"]`
+      );
+      if (el) {
+        const button = el as HTMLButtonElement;
+        params.set(button.name, button.value);
+        loadingEvent.detail.button = button;
+      }
+    }
+
+    this.element.dispatchEvent(loadingEvent);
+    try {
+      const resp = await axios(form.action, {
+        method: form.method as Method,
+        headers: {
+          "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
+          "X-Authgear-XHR": "true",
+        },
+        data: params,
+        onUploadProgress: progressEventHandler,
+        onDownloadProgress: progressEventHandler,
+      });
+
+      const { redirect_uri, action } = resp.data;
+
+      clearCache();
+      switch (action) {
+        case "redirect":
+          // Perform full redirect.
+          window.location = redirect_uri;
+          break;
+
+        case "replace":
+        case "advance":
+          // We assume the URL returned by the server have query preserved.
+          visit(redirect_uri, { action });
+          break;
+      }
+    } catch (e: unknown) {
+      handleAxiosError(e);
+      this.element.dispatchEvent(
+        new CustomEvent("onLoadingError", {
+          detail: {
+            error: e,
+          },
+        })
+      );
+    } finally {
+      this.element.dispatchEvent(new CustomEvent("onLoadingEnd"));
+    }
+  }
+
+  connect() {
+    const elems = document.querySelectorAll("form");
+    for (let i = 0; i < elems.length; i++) {
+      if (elems[i].querySelector('[data-turbo="false"]')) {
+        continue;
+      }
+      this.forms.push(elems[i] as HTMLFormElement);
+    }
+    for (const form of this.forms) {
+      form.addEventListener("submit", this.onSubmitCapture, true);
+      form.addEventListener("submit", this.onSubmit);
+    }
+  }
```

Comment:
The original implementation listened to events of turbo, I guess they are still needed?

Reply:
Oh I found them in loading controller.